File: src/60.Permutation_Sequence-Math/permutation_sequence.cpp
#include<iostream>
#include<string>
#include<vector>
using namespace std;

class Solution {
public:
    string getPermutation(int n, int k) {
        vector<int> board(10,1);
        vector<int> ilabel(n+1,0);
        string res;
        n = n;
        k = k - 1;
        
        for(int i = 1; i < 10; i++){
            board[i] = board[i-1] * i;
        }

        while(n){
            int bit = k / board[n-1];
            k = k % board[n-1];
            int flag = 0;
            int i;
            for(i = 1; flag <= bit; i++){
                if(ilabel[i] == 0)flag++;
            }
            res = res + to_string(i-1);
            ilabel[i-1] = 1;
            n--;
        }
        return res;    
    }
};

int main()
{
    Solution S;
    int n = 3;
    int k = 4;
    string res = S.getPermutation(n, k);
    cout << res << endl;
    return 0;
}

File: src/0060.Permutation_Sequence-Math/permutation_sequence.java
// 回溯法 + 剪枝
class Solution {
    boolean[] used;
    int[] factorial;
    int n;
    int k;
    public String getPermutation(int n, int k) {
        this.n = n;
        this.k = k;
        used = new boolean[n+1];
        factorial = new int[n+1];
        factorial[0] = 1;
        for (int i = 1; i <= n; i++) {
            factorial[i] = factorial[i-1] * i;
        }
        StringBuilder path = new StringBuilder();
        dfs(0, path);
        return path.toString();
    }
    private void dfs(int index, StringBuilder sb) {
        if (index == n) {
            return ;
        }
        int res = factorial[n-index-1];
        for (int i = 1; i <= n; i++) {
            if (used[i]) {
                continue;
            }
            if (k > res) {
                k -= res;
                continue;
            }
            sb.append(i);
            used[i] = true;
            dfs(index+1, sb);
            return ;
        }
    }
}